### PR TITLE
feat: format the solidity test output better

### PR DESCRIPTION
**Pull request by galargh**
- [ ] Because this PR includes a **bug fix**, relevant tests have been included.
- [ ] Because this PR includes a **new feature**, the change was previously discussed on an Issue or with someone from the team.
- [x] I didn't do anything of this.

---

This PR builds up on https://github.com/NomicFoundation/hardhat/pull/5809 and introduces custom output formatting described in https://www.notion.so/nomicfoundation/Reporter-Output-Formatting-11f578cdeaf580408d12fb03e7799e81.

It has been tested manually and visually verified.

The the reporter produces for the example project after the changes:
Screenshot 2024-10-16 at 12 11 03

### CODE REVIEW

File: v-next/hardhat/src/internal/builtin-plugins/solidity-test/reporter.ts [hunk 1]
```diff
@@ -14,47 +19,119 @@ import chalk from "chalk";
...
       case "suite:result": {
         const { data: suiteResult } = event;
-        for (const testResult of suiteResult.testResults) {
-          testResultCount++;
-          let name = suiteResult.id.name + " | " + testResult.name;
-          if ("runs" in testResult?.kind) {
-            name += ` (${testResult.kind.runs} runs)`;
+        const suiteTestCount = suiteResult.testResults.length;
+
+        if (suiteTestCount === 0) {
```
Comment (alcuadrado) [inline v-next/hardhat/src/internal/builtin-plugins/solidity-test/reporter.ts:36]: When can this happen? An empty test contract, for example?

File: v-next/hardhat/src/internal/builtin-plugins/solidity-test/reporter.ts [hunk 2]
```diff
@@ -14,47 +19,119 @@ import chalk from "chalk";
...
+        // NOTE: The test results are in reverse run order, so we reverse them
+        // again to display them in the correct order.
+        for (const testResult of suiteResult.testResults.reverse()) {
+          const name = testResult.name;
           const status: TestStatus = testResult.status;
+          const details = [
+            ["duration", `${testResult.durationMs} ms`],
+            ...Object.entries(testResult.kind),
+          ]
+            .map(([key, value]) => `${key}: ${value}`)
```
Comment (alcuadrado) [inline v-next/hardhat/src/internal/builtin-plugins/solidity-test/reporter.ts:62]: When `value` is a `Buffer`, we should format it as a 0x-prefixed hex string.

Comment (alcuadrado) [reply v-next/hardhat/src/internal/builtin-plugins/solidity-test/reporter.ts:62]: I think this can also happen later in this code.

**Review by alcuadrado — APPROVED** [review]
Left a minor comment, but LGTM

File: v-next/hardhat/src/internal/builtin-plugins/solidity-test/reporter.ts (hunk 1 above)
Comment (galargh) [reply v-next/hardhat/src/internal/builtin-plugins/solidity-test/reporter.ts:36]: Yes, exactly or an abstract one only.

File: v-next/hardhat/src/internal/builtin-plugins/solidity-test/reporter.ts (hunk 2 above)
Comment (galargh) [reply v-next/hardhat/src/internal/builtin-plugins/solidity-test/reporter.ts:62]: Got it! I'll introduce a utility `bufferToHex` copied from v2 and use it here in a follow-up PR.

Comment (galargh) [reply v-next/hardhat/src/internal/builtin-plugins/solidity-test/reporter.ts:62]: Here is the follow-up PR - https://github.com/NomicFoundation/hardhat/pull/5847

**Pull request merged**